### PR TITLE
SN32: Add PAL callback support

### PR DESCRIPTION
**Pull request by 1Conan**
Adds PAL callbacks to SN32F2xx

Tested on SN32F248B

### CODE REVIEW

**Review by dexter93 — DISMISSED** [review]

File: os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c [hunk 1]
```diff
@@ -54,21 +54,18 @@ static void initgpio(ioportid_t gpiop, const sn32_gpio_setup_t *config) {
...
 
 #if (PAL_USE_WAIT == TRUE) || (PAL_USE_CALLBACKS == TRUE)
-static void irq_handler(ioportid_t port, palevent_t *events) {
+static void irq_handler(ioportid_t port, uint32_t offset) {
     chSysLockFromISR();
     uint32_t ris = port->RIS; /* Get pending interrupts on this port. */
     port->IC = 0xFFFFFFFFU; /* Clear all pending interrupts on this port. */
     chSysUnlockFromISR();
 
     iopadid_t pad;
```
Comment (dexter93) [inline os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c]:
```suggestion
```

File: os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c [hunk 2]
```diff
@@ -54,21 +54,18 @@ static void initgpio(ioportid_t gpiop, const sn32_gpio_setup_t *config) {
...
 
     /* invoke callbacks for pending interrupts */
-    for (pad = 0, pad_mask = 1U;
-         pad < PAL_IOPORTS_WIDTH;
-         ++pad, ++events, pad_mask <<= 1) {
-        if (events->cb && (ris & pad_mask)) {
-            events->cb(events->arg);
+    for (iopadid_t pad = 0; pad < PAL_IOPORTS_WIDTH; pad++) {
+        if (ris & (1 << pad)) {
+            _pal_isr_code(offset + pad)
```
Comment (dexter93) [inline os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c]:
```suggestion
            _pal_isr_code(offset + pad);
```

File: os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c [hunk 3]
```diff
@@ -294,6 +291,16 @@ void _pal_lld_enablepadevent(ioportid_t ioport,
  */
 void _pal_lld_disablepadevent(ioportid_t ioport, iopadid_t pad) {
   _pal_lld_enablepadevent(ioport, pad, PAL_EVENT_MODE_DISABLED);
```
Comment (dexter93) [inline os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c]:
```suggestion
void _pal_lld_disablepadevent(ioportid_t port, iopadid_t pad) {
  _pal_lld_enablepadevent(port, pad, PAL_EVENT_MODE_DISABLED);
```

**Review by dexter93 — CHANGES_REQUESTED** [review]
changes to make it build

File: os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c [hunk 4]
```diff
@@ -142,6 +217,104 @@ void _pal_lld_setpadmode(ioportid_t port,
...
+/**
+ * @brief   Enables a pad event.
+ *
+ * @param[in] port   port containing pad whose event is to be enabled
+ * @param[in] pad    pad whose event is to be enabled
+ * @param[in] mode   mode of the event
+ *
+ * @notapi
+ */
+void _pal_lld_enablepadevent(ioportid_t ioport,
```
Comment (dexter93) [inline os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c]:
```suggestion
void _pal_lld_enablepadevent(ioportid_t port,
```

File: os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c [hunk 5]
```diff
@@ -142,6 +217,104 @@ void _pal_lld_setpadmode(ioportid_t port,
...
+                             iopadid_t pad,
+                             iomode_t mode) {
+  switch(mode) {
+    case PAL_EVENT_MODE_RISING_EDGE:
+      ioport->IS  &= ~(1 << pad);
+      ioport->IBS &= ~(1 << pad);
+      ioport->IEV &= ~(1 << pad);
+      ioport->IE  |= (1 << pad);
```
Comment (dexter93) [inline os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c]:
```suggestion
      port->IS  &= ~(1 << pad);
      port->IBS &= ~(1 << pad);
      port->IEV &= ~(1 << pad);
      port->IE  |= (1 << pad);
```

File: os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c [hunk 6]
```diff
@@ -142,6 +217,104 @@ void _pal_lld_setpadmode(ioportid_t port,
...
+      break;
+    case PAL_EVENT_MODE_FALLING_EDGE:
+      ioport->IS  &= ~(1 << pad);
+      ioport->IBS &= ~(1 << pad);
+      ioport->IEV |= (1 << pad);
+      ioport->IE  |= (1 << pad);
```
Comment (dexter93) [inline os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c]:
```suggestion
      port->IS  &= ~(1 << pad);
      port->IBS &= ~(1 << pad);
      port->IEV |= (1 << pad);
      port->IE  |= (1 << pad);
```

File: os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c [hunk 7]
```diff
@@ -142,6 +217,104 @@ void _pal_lld_setpadmode(ioportid_t port,
...
+      break;
+    case PAL_EVENT_MODE_BOTH_EDGES:
+      ioport->IS  &= ~(1 << pad);
+      ioport->IBS |= (1 << pad);
+      ioport->IE  |= (1 << pad);
```
Comment (dexter93) [inline os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c]:
```suggestion
      port->IS  &= ~(1 << pad);
      port->IBS |= (1 << pad);
      port->IE  |= (1 << pad);
```

File: os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c [hunk 8]
```diff
@@ -142,6 +217,104 @@ void _pal_lld_setpadmode(ioportid_t port,
...
+      break;
+    case PAL_EVENT_MODE_DISABLED:
+      ioport->IS  &= ~(1 << pad);
+      ioport->IBS &= ~(1 << pad);
+      ioport->IEV &= ~(1 << pad);
+      ioport->IE  &= ~(1 << pad);
```
Comment (dexter93) [inline os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c]:
```suggestion
      port->IS  &= ~(1 << pad);
      port->IBS &= ~(1 << pad);
      port->IEV &= ~(1 << pad);
      port->IE  &= ~(1 << pad);
```

File: os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c [hunk 9]
```diff
@@ -142,6 +217,104 @@ void _pal_lld_setpadmode(ioportid_t port,
...
+      break;
+    default:
+      return;
+  }
+}
+
+/**
+ * @brief   Disables a pad event
+ *
+ * @param[in] ioport    port containing pad whose event is to be disabled
```
Comment (dexter93) [inline os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c]:
```suggestion
 * @param[in] port    port containing pad whose event is to be disabled
```

File: os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c [hunk 10]
```diff
@@ -142,6 +217,104 @@ void _pal_lld_setpadmode(ioportid_t port,
...
+ * @param[in] pad       pad whose event is to be disabled
+ *
+ * @notapi
+ */
+void _pal_lld_disablepadevent(ioportid_t ioport, iopadid_t pad) {
+  _pal_lld_enablepadevent(ioport, pad, PAL_EVENT_MODE_DISABLED);
```
Comment (dexter93) [inline os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c]:
```suggestion
void _pal_lld_disablepadevent(ioportid_t port, iopadid_t pad) {
  _pal_lld_enablepadevent(port, pad, PAL_EVENT_MODE_DISABLED);
```

File: os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c [hunk 11]
```diff
@@ -142,6 +217,104 @@ void _pal_lld_setpadmode(ioportid_t port,
...
+    _pal_clear_event(PAL_IOPORTS_WIDTH * 2 + pad);
+  } else if (port == GPIOD) {
+    _pal_clear_event(PAL_IOPORTS_WIDTH * 3 + pad);
+  }
+}
+
+/**
+ * @brief   Returns whether a pad event is enabled
+ *
+ * @param[in] ioport   port containing the pad
```
Comment (dexter93) [inline os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c]:
```suggestion
 * @param[in] port   port containing the pad
```

File: os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c [hunk 12]
```diff
@@ -142,6 +217,104 @@ void _pal_lld_setpadmode(ioportid_t port,
...
+ * @param[in] pad      pad whose event is to be queried
+ *
+ * @notapi
+ */
+bool _pal_lld_ispadeventenabled(ioportid_t ioport, iopadid_t pad) {
+  return ((ioport->IE >> pad) & 1);
```
Comment (dexter93) [inline os/hal/ports/SN32/LLD/SN32F2xx/GPIO/hal_pal_lld.c]:
```suggestion
bool _pal_lld_ispadeventenabled(ioportid_t port, iopadid_t pad) {
  return ((port->IE >> pad) & 1);
```

**Review by dexter93 — CHANGES_REQUESTED** [review]
standardize on `port` usage vs `ioport` across the pal stack

**Review by fpoussin — APPROVED** [review]

**Comment by fpoussin** [comment]
Thank you :)

**Pull request merged**